### ISSUE TITLE
@codingStandardsIgnoreStart does not work

### ISSUE THREAD

**Issue by prashantadevkota** [squizlabs/PHP_CodeSniffer#1258]
Ok, first of all, I've gone through all previous bug reports related with this issue, and unfortunately, I have the same problem.

I am using CodeIgniter, which places a defined('BASEPATH') in every file. This means that the PSR2 rule will get violated for each and every file. Excluding a particular file from sniffing is NOT an option for me.

// @codingStandardsIgnoreStart
defined('BASEPATH') OR exit('No direct script access allowed');
// @codingStandardsIgnoreEnd

class SomeClass
{
  .....
}

I still get the "A file should declare new symbols (classes, functions, constants, etc.) and cause no other side effects, or it should execute logic with side effects, but should not do both." error.

I even tried "// @codingStandardsIgnoreLine" to no avail.

**Comment by gsherwood** [comment]
Those comments don't stop sniffs from processing the lines that they surround. They just mute errors that are produced on those lines.

The particular error message you are getting would be on line 1, because that sniff is doing a general report of the file and not specific lines in the file. So those comments aren't going to help you with that sniff.

What you need to do is create a custom coding standard (just an XML file) and exclude that sniff from the PSR2 standard. Then tell your plugin to use your custom standard and not the built-in PSR2 standard.

Your custom coding standard will be an XML file (normally called ruleset.xml or phpcs.xml) and have the following content:
```xml
<?xml version="1.0"?>
<ruleset name="MyStandard">
    <description>My custom coding standard.</description>
    <rule ref="PSR2">
        <exclude name="PSR1.Files.SideEffects"/>
    </rule>
</ruleset>
```
Instead of specifying the coding standard name as "PSR2", specify the path to that XML file instead. If you were running from the command line, you'd use this command: `phpcs --standard=/path/to/ruleset.xml /path/to/code`

Does that explain things?

**Comment by prashantadevkota** [comment]
First of all thank you for your quick feedback. What you have mentioned is exactly what I am doing at the moment. I have excluded PSR1.Files.SideEffects and PSR1.Classes.ClassDeclaration (because CodeIgniter), and yes it does solve this particular problem.

So I guess this could go as a desired feature in PHPCS then? Because I would still like the PSR rule to be enforced for other files/areas.

**Comment by gsherwood** [comment]
> So I guess this could go as a desired feature in PHPCS then? Because I would still like the PSR rule to be enforced for other files/areas.

Yes, although I'm not sure how I'd do that specifically.

You can apply exclusions only to specific files, folders, or use a regex to pattern match the file paths. The docs for that are:
```xml
 <!--
    You can hard-code ignore patterns for specific sniffs,
    a feature not available on the command line. Please note that
    all sniff-specific ignore patterns are checked using absolute paths.

    The code here will hide all messages from the Squiz DoubleQuoteUsage
    sniff for files that match either of the two exclude patterns.
 -->
 <rule ref="Squiz.Strings.DoubleQuoteUsage">
    <exclude-pattern>*/tests/*</exclude-pattern>
    <exclude-pattern>*/data/*</exclude-pattern>
 </rule>
```

I'm not sure if that is an option for you in this case, but it might be worth exploring.

**Comment by gsherwood** [comment]
From version 3.2, you'll be able to exclude specific code from being checked by the SideEffects sniff. E.g.,
```php
// phpcs:disable PSR1.Files.SideEffects
defined('BASEPATH') OR exit('No direct script access allowed');
// phpcs:enable PSR1.Files.SideEffects
```
Or if you don't want the line checked at all:
```php
// phpcs:disable
defined('BASEPATH') OR exit('No direct script access allowed');
// phpcs:enable
```

See issue #604 for more details.
